Implement configuration mechanism

radoslav-d opened the pull request.

radoslav-d committed:
Implement configurable service broker

Add rest points for configuring the behavior of the budgie service
broker. Create a schema for defining the behavior of the service broker
and mainly the fail behavior. It is possible to configure whether the
broker operates synchronous or asynchronous and the timeout.

radoslav-d committed:
Implement async operations and bindings

Introduce a logic, which handles asynchronous operations. Add rest end
points for getting the last operation status (from the spec). Add
support for binding and unbinding. Add a rest end point for adding new
service offerings.

radoslav-d committed:
Fix bugs in async operations and bindings

radoslav-d committed:
Refactor BehaviorConfiguration schema and introduce validation

radoslav-d committed:
Implement configuration to path param mapping

Extend the rest api to accept additional path param, which will be used
to distinguish between different configurations. In that way, a service
broker can be registered to different route resulting in possibility for
configuring the behavior exactly for this route.

enchobelezirev commented on README.md:
````diff
@@ -0,0 +1,117 @@
+
+```
+{
+    "async": 5000,
````
Does this apply to the above text? the async? Shouldn't be asyncDuration?

enchobelezirev commented on manifest.yml:
```diff
@@ -2,10 +2,10 @@
   - name: budgie-sb
     host: budgie-sb
-    path: target/budgie-sb-0.0.1-SNAPSHOT.war
+    path: target/budgie-sb-0.1.0-SNAPSHOT.war
```
Could we release this and not use SNAPSHOT ?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/BehaviorEngine.java:
```diff
@@ -0,0 +1,155 @@
+
+    private ApplicationConfiguration appConfiguration;
+
+    private Map<String, BehaviorConfiguration> configurations = new HashMap<>();
```
Why not use synchronized map and remove the synchronization from the methods?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/BehaviorEngine.java:
```diff
@@ -0,0 +1,155 @@
+import java.util.stream.Stream;
+
+@Component
+public class BehaviorEngine {
```
*Engine... do not go in danger waters with names - be careful what names you put to your classes- try to be as much as possible explicit but do not make it too abstract. From this naming I could assume that this class is doing some heavy processing and complex analysis ...

radoslav-d replied:
Does `ConfigurationManager` sound good?

enchobelezirev replied:
It Manager is the same as the Engine :D

radoslav-d replied:
:D do you have any name in mind?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/BehaviorConfiguration.java:
```diff
@@ -0,0 +1,44 @@
+
+public class BehaviorConfiguration {
+
+    private UUID id;
```
What is this used for?

enchobelezirev replied:
Could you introduce better naming? and also for the class?

radoslav-d replied:
It is not needed anymore, I will remove it.

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/BehaviorEngine.java:
```diff
@@ -0,0 +1,155 @@
+        if (isAsync(id)) {
+            return configuration.getAsyncDuration();
+        }
+        if (configuration != null && configuration.getSyncDuration() != null) {
```
Could we make isSync method which is consistent with the other methods isAsync method?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/BehaviorEngine.java:
```diff
@@ -0,0 +1,155 @@
+        if (configuration != null && configuration.getSyncDuration() != null) {
+            return configuration.getSyncDuration();
+        }
+        return 0;
```
Is this the default duration? Could this be problematic? For instance, I forget to specify duration, what will happen then?

radoslav-d replied:
Nothing will happen. I mean that the async operations will be still async but with no downtime and the sync operations will not sleep.

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/BehaviorEngine.java:
```diff
@@ -0,0 +1,155 @@
+        }
+        return configuration.getFailConfigurations().stream()
+            .filter(failConfig -> operationType.equals(failConfig.getOperationType()))
+            .filter(failConfig -> all().or(byInstanceId()
```
all, and, by*... are not understandable. If you would like you can check StepBuilder pattern and introduce nice reading for the criteria building logic

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/FailConfiguration.java:
```diff
@@ -0,0 +1,101 @@
+        return all;
+    }
+
+    public void setAll(Boolean all) {
```
What is this all?

radoslav-d replied:
will rename it

enchobelezirev commented on src/main/java/com/sap/broker/budgie/configuration/behavior/FailConfiguration.java:
```diff
@@ -0,0 +1,101 @@
+import java.util.List;
+import java.util.UUID;
+
+public class FailConfiguration {
```
You can use Builder to remove the setters and make the class Immutable?
Or you do not need it to be immutable?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/domain/Binding.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Map;
+import java.util.UUID;
+
+public class Binding {
```
Is this really binding? Is that service instance or just service?
Binding is something that represents the connection between given application and a service.

radoslav-d replied:
The purpose of this class is to hold the meta information about the binding. The bindings this service broker provides are dummy ones.

enchobelezirev replied:
Yes but again, the Binding you represent here is not a real one - even though you are using this class as a Mocked class.

enchobelezirev commented on src/main/java/com/sap/broker/budgie/domain/ServiceInstance.java:
```diff
@@ -53,6 +55,36 @@ public void setParameters(Map<String, Object> parameters) {
+
+    private void validateBindingMap() {
+        if (bindings == null) {
+            setBindings(new HashMap<>());
```
You could eliminate this by putting it in the class declaration

radoslav-d replied:
I do not think it is a good idea because this object flies through http and the serialization process does not instantiate the map.

enchobelezirev commented on src/main/java/com/sap/broker/budgie/helpers/AsyncOperation.java:
```diff
@@ -0,0 +1,33 @@
+    public AsyncOperation(Supplier<AsyncOperationState> operation, int milliseconds) {
+        this.operation = operation;
+        this.milliseconds = milliseconds;
+        new Thread(this).start();
```
I think this is not cool.
Could you extract it outside the class and make some sort of executor for this? Or even better, introduce ExecutorService?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/exception/InterruptedOperationException.java:
```diff
@@ -0,0 +1,19 @@
+package com.sap.broker.budgie.exception;
+
+public class InterruptedOperationException extends RuntimeException {
```
Why do we need special type of exception?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/impl/ServiceBroker.java:
```diff
@@ -30,35 +30,35 @@ public Catalog getCatalog() {
     }
 
-    public Collection<ServiceInstance> getAll() {
+    public synchronized Collection<ServiceInstance> getAll() {
```
Same question about the syncronized map

enchobelezirev commented on src/main/java/com/sap/broker/budgie/helpers/BehaviorConfigurationValidator.java:
```diff
@@ -0,0 +1,38 @@
+        if (configuration.getFailConfigurations() != null) {
+            return configuration.getFailConfigurations().stream().allMatch(config -> {
+                Integer status = config.getStatus();
+                return status != null && status >= 400 && status < 600;
```
Extract this into method and apply this for all the lambdas that are in use.

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/CatalogResource.java:
```diff
@@ -1,28 +1,46 @@
-@Path("/catalog")
+import java.util.List;
+
+@Path("/{configId}/v2/catalog")
```
Isnt't that the spaceId and not configId?

radoslav-d replied:
the space name is used in the OQ tests because of its uniqueness across the tests

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ConfigurationResource.java:
```diff
@@ -0,0 +1,59 @@
+    private BehaviorConfigurationValidator behaviorConfigurationValidator;
+
+    @Inject
+    public ConfigurationResource(BehaviorEngine behaviorEngine, BehaviorConfigurationValidator behaviorConfigurationValidator) {
```
Nice. What are the benefits of this? :)

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ConfigurationResource.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @GET
+    public Response getAll() {
```
getAll what?

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ConfigurationResource.java:
```diff
@@ -0,0 +1,59 @@
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+@Path("/v2/configuration")
```
not consistent with the other APIs:
/{configId}/v2/....

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ServiceInstancesResource.java:
```diff
@@ -38,69 +41,274 @@ public ServiceInstancesResource(ServiceBroker serviceBroker) {
+    public Response create(@PathParam("configId") String configId, @PathParam("id") UUID id, @QueryParam("accepts_incomplete") boolean acceptIncomplete, ServiceInstance serviceInstance) {
+        if (behaviorEngine.isAsync(configId)) {
+            if (!acceptIncomplete) {
+                return emptyBodyResponse(422);
```
'422' use HtttpStatuseCode...

radoslav-d replied:
422 is not available in `javax.ws.rs.core.Response.Status`

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ServiceInstancesResource.java:
```diff
@@ -38,69 +41,274 @@ public ServiceInstancesResource(ServiceBroker serviceBroker) {
 
     @GET
     @Path("/{id}")
```
I would put better naming than just id. Like serviceInstanceGuid

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ServiceInstancesResource.java:
```diff
@@ -38,69 +41,274 @@ public ServiceInstancesResource(ServiceBroker serviceBroker) {
+    }
+
+    private Response createSync(String configId, UUID id, ServiceInstance serviceInstance) {
+        sleepIfWanted(configId);
```
rename to sleepForDurationTimeout or something like that.

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ServiceInstancesResource.java:
```diff
@@ -38,69 +41,274 @@ public ServiceInstancesResource(ServiceBroker serviceBroker) {
+    }
+
+    private Response emptyBodyResponse(int httpStatus) {
+        return Response.status(httpStatus).entity(new Object()).build();
```
you can extract new Object() into constant string and name it EMPTY_RESPONSE_BODY

enchobelezirev commented on src/main/java/com/sap/broker/budgie/resources/api/ServiceInstancesResource.java:
```diff
@@ -38,69 +41,274 @@ public ServiceInstancesResource(ServiceBroker serviceBroker) {
     }
 
+    private Response createAsync(String configId, UUID id, ServiceInstance serviceInstance) {
+        AsyncOperation asyncOperation = new AsyncOperation(() -> {
```
extract this runnable creation into another method for better reading
Even better, you would be able to reuse  the newly extracted method

enchobelezirev commented:
Finally, could you show me example Requests and Responses in order to get the full picture of the usage of the buggie service broker?

radoslav-d commented:
> Finally, could you show me example Requests and Responses in order to get the full picture of the usage of the buggie service broker?

![Screen Shot 2020-02-13 at 15 19 36](https://user-images.githubusercontent.com/32434758/74439131-49d39f80-4e74-11ea-9aa0-a846ec7b956a.png)


![Screen Shot 2020-02-13 at 15 21 47](https://user-images.githubusercontent.com/32434758/74439251-915a2b80-4e74-11ea-8192-1aa2eca04ce9.png)

radoslav-d committed:
Change the api routes to be more self-explanatory

Add validation for configurations, which check if service ids and names
as well as plan ids and names exist.
